esp-hal/src/i2c/master/mod.rs

```diff
@@ -85,6 +86,107 @@ const I2C_CHUNK_SIZE: usize = 254;
 // on ESP32 there is a chance to get trapped in `wait_for_completion` forever
 const MAX_ITERATIONS: u32 = 1_000_000;
 
+/// Address mode.
+///
+/// Note: This trait is sealed and should not be implemented outside of this
+/// crate.
+pub trait AddressMode: _private::AddressModeInternal + Copy + 'static {}
+
+mod _private {
+    pub trait AddressModeInternal {
+        fn as_i2c_address(&self) -> super::I2cAddress;
+    }
+
+    /// Internally used representation of an I2C address.
+    #[derive(Debug, Copy, Clone, PartialEq, Eq, Hash, strum::Display)]
+    #[cfg_attr(feature = "defmt", derive(defmt::Format))]
+    pub enum I2cAddress {
+        /// Seven bit address
+        SevenBit(super::SevenBitAddress),
+    }
+
+    impl I2cAddress {
+        pub fn as_u8(&self) -> u8 {
+            match self {
+                I2cAddress::SevenBit(addr) => *addr,
+            }
+        }
+    }
+}
+
+/// 7-bit address mode type.
+///
+/// Note that 7-bit addresses defined by drivers should be specified in
+/// **right-aligned** form, e.g. in the range `0x00..=0x7F`.
+///
+/// For example, a device that has the seven bit address of `0b011_0010`, and
+/// therefore is addressed on the wire using:
+///
+/// * `0b0110010_0` or `0x64` for *writes*
+/// * `0b0110010_1` or `0x65` for *reads*
+pub type SevenBitAddress = u8;
+
+impl AddressMode for SevenBitAddress {}
+
+impl _private::AddressModeInternal for SevenBitAddress {
+    fn as_i2c_address(&self) -> I2cAddress {
+        I2cAddress::SevenBit(*self)
+    }
+}
+
+/// I2C SCL timeout period.
+///
+/// When the level of SCL remains unchanged for more than `timeout` bus
+/// clock cycles, the bus goes to idle state.
+///
+/// Default value is `BusCycles(10)`.
+#[doc = ""]
+#[cfg_attr(
+    not(esp32),
+    doc = "Note that the effective timeout may be longer than the value configured here."
+)]
+#[derive(Debug, Copy, Clone, PartialEq, Eq, Hash, strum::Display)]
+#[cfg_attr(feature = "defmt", derive(defmt::Format))]
+#[non_exhaustive]
+// TODO: when supporting interrupts, document that SCL = high also triggers an
+// interrupt.
+pub enum Timeout {
+    /// Use the maximum timeout value.
+    Maximum,
+
+    /// Disable timeout control.
+    #[cfg(not(any(esp32, esp32s2)))]
+    Disabled,
+
+    /// Timeout in bus clock cycles.
+    BusCycles(u32),
+}
+
+impl Timeout {
+    fn cycles(&self) -> u32 {
+        match self {
+            #[cfg(esp32)]
```

Comment:
nitpick: Shouldn't those be `cfg_if!` (or `cfg!`) regarding the APIP GUIDELINES?

Reply:
IMHO it looks worse in this case but happy to change is requested